### ISSUE TITLE
Flag invalid calculated coordinates

### ISSUE THREAD

**Issue by pstorch** [cgeo/cgeo#6697]
##### Detailed steps causing the problem:
- Add a waypoint with calculated coordinates
- Enter some formulas/values which result in errors
- The preview below the variables show, that there is something wrong, either with an '*' or with a negative value
- Save the current state and go back to the waypoint

##### Actual behavior after performing these steps:
- Valid coordinates are shown (some digits are 0 but valid)
- There is no indication in the waypoint or in the waypoint list that there is something wrong with the coordinate

##### Expected behavior after performing these steps:
- Flag the waypoint as errornous

##### Version of c:geo used:
- nightly

##### Is the problem reproducible for you?
Yes

##### Other comments and remarks:
Errors in the calculator can be missed easily. When changing a variable value the keyboard is overlayed and the coordinate result is not visible. If the user just clicks the checkmark to close the dialog without first closing the keyboard to check the result, an error is missed. It's not visible in the other dialogs, that this waypoint has a problem.

**Comment by S-Bartfast** [comment]
Yes, the error detection most certainly leaves a bit to be desired.

As it is I simply lifted the exact same checking algorithm as used in the regular coordinate entry validation.  I tried to follow what was going on inside that checking algorithm a few times but got so thoroughly lost I just threw my hands up and assumed "If it's good enough for them..."

While it leaves a lot to be desired try entering the exact same text into the 'plain' format of the regular coordinate input and see if it's accepted by that parser.  If you're getting different behavior I'll look into it more closely but if it's the same than I dare say it's a deeper issue with parsing in general and not specifically related to the calculator as such.

**Comment by S-Bartfast** [comment]
Commit [78e84cf489e51200adf4eb89c1e998b7827ff774](https://github.com/cgeo/cgeo/commit/78e84cf489e51200adf4eb89c1e998b7827ff774) (Implemented showing a "Save" icon to replace the "tick" icon for the "Done" button when the resulting coordinates are invalid) should have gone some of the way to resolving this issue.

Also, as part of the work for #6716 the asterisk ("*") and underscore ("_") characters are now considered invalid characters so far as the calculator is concerned.

Do these fixes address the concerns of this ticket?

**Comment by pstorch** [comment]
Not really. As long as the calculator can't finish its job, the waypoint has "some other" coordinates.
I don't see in the waypoint list, that these coordinates are not "ready". 

We have waypoints without coordinates. Maybe it's better to not show coordinates as long as they are not completely calculated.

**Comment by S-Bartfast** [comment]
> We have waypoints without coordinates. Maybe it's better to not show coordinates as long as they are not completely calculated.

Hmm.
So at the moment, if you enter partial coordinates into the calculator and click "save" the calculators state will be stored in that waypoints data.  If you then click "cancel" (or hit back twice) you'll be taken back to the "Add waypoint" activity where you can save the waypoint (which must be done if you want to keep the calculators state) and it will be saved with the coordinates shown in the page.  These will be the coordinates originally shown in the activity as "cancel" was clicked to exit the calculator (rather than because the calculator's coordinates were invalid).  Note the same would happen if you where to click "cancel" even if you did have valid coordinates in the calculator.  It is the "done" (tick button) that updates the coordinates of the waypoint to those shown in the calculator and this is exactly the same as how the "Coordinate Input" dialog works.  There is an added complexity in that the calculator has state one may want to preserve but it's the fact that the "done" (tick) button is never clicked that the coordinates are not updated rather than because the partial coordinates are invalid as such.

It would be possible to have clicking the "save" button do something to the underlying GeoPoint and put it into some kind of 'flagged' state but this would mean it would no longer be possible to "cancel" the calculator and return to the previous coordinates, which brings us back to the #6696 discussion.

**Comment by Lineflyer** [comment]
> Not really. As long as the calculator can't finish its job, the waypoint has "some other" coordinates.
I don't see in the waypoint list, that these coordinates are not "ready".

> We have waypoints without coordinates. Maybe it's better to not show coordinates as long as they are not completely calculated.

While using the RC-version for a long cache trip yesterday I had the same thinking.
Before starting on my cache tour I created final waypoints for the multi caches to be solved and entered the needed formulas. So far so good.
Out in the field I was a little confused by a lot of final waypoint icons showing up in parallel to the cache icons or at my home coordinates (as by default either the cache location or the current location is used when creating the WP).

So I agree, at least if the calc result is invalid the corresponding waypoint coordinated should be NULL/empty.

**Comment by Matze2** [comment]
Also here a 👍 . Having (quite arbitrary) waypoint coordinates in case of invalid or better incomplete calculator input is quite confusing.

**Comment by S-Bartfast** [comment]
So you're advocating that clicking 'Save' in the calculator should render the waypoints coordinates invalid, yeah?

This can be implemented easy enough but it of course means that you will then not be able to 'Cancel' the calculator and return to the way the waypoint was previously, just so were clear.

**Comment by Matze2** [comment]
My feeling says that it is a problem that we allow to edit waypoint coordinates directly and in parallel with the calculator. In real life, it is probably an either-or.
As soon as I entered my formulas in the calculator I would expect that everything what I do there is also reflected in the waypoint coordinates. If I have all variables filled and press "Save" then the coordinates are also overwritten. Why not when I do not have valid coordinates? So for me that sounds natural.
Maybe it is a problem that "Save" does two things at once: saving the calculator data (formulas, variables values) and overwriting the coordinate.

**Comment by S-Bartfast** [comment]
At the moment it is only the "done" button ('tick' icon) that overwrites the waypoint's coordinates.  The "save" button ('disk' icon) only saves the calculator's state.  This can of course be changed but we will have to come up with a 'flag' value of some sort for invalid coordinates because at present we don't have a notion of a waypoint with invalid coordinates.

For the Coordinate Input Dialog, if invalid coordinates are entered the Coordinate Input Dialog simply refuses to close (unless you cancel it).  What we are wanting to do with the calculator is "save" invalid coordinates which is something we've never dealt with before.  That's not to say we can't work out a way (I mean heck, we have an international team of dedicated engineers working on this ;)).

**Comment by pstorch** [comment]
> at present we don't have a notion of a waypoint with invalid coordinates.

But latitude and longitude of a waypoint can be NULL. If the waypoint is a calculated waypoint I think NULL could be a good indicator, that there is no valid result yet of the calculation. Wdyt?

**Comment by S-Bartfast** [comment]
>But latitude and longitude of a waypoint can be NULL.

Are you sure?
I did try it and 'EditWaypointActivty' was crashing when 'updateCoordinates(...)' was called, but even fixing that it just doesn't seem a way point can be saved without a valid GeoPoint:
![image](https://user-images.githubusercontent.com/6573522/32440832-d24bcbf2-c344-11e7-872d-4a6c849d5b9f.png)

**Comment by pstorch** [comment]
Sorry my bad. I was only talking about the database. There we have for sure waypoints with lat/lon = NULL. The `updateCoordinates(...)` method doesn't support it (yet).
Look at http://coord.info/GC768Y3 for example. There are some stages without coordinates.

**Comment by S-Bartfast** [comment]
Fair enough.

Just to be clear, it's not only the 'updateCoordinates(...)' method that was the problem.  The 'updateCoordinates(...)' was causing a crash but even fixing that the 'EditWaypointActivity' itself won't add a waypoint if the coordinates are not valid. The issue is that we end up in the 'catch-block' on line 548 of 'EditWaypointActivity.java' when we attempt to save the waypoint.

**Comment by Matze2** [comment]
This means before this issue can be resolved we need to resolve the issue "can not clear coordinates in a waypoint".
And this is a real issue. Let's take GC768Y3 again as example. You want to add a user note to stage 18 which does not have any coordinates yet. As soon as  saved the waypoint you just get the coordinates of your current position. This also confused me in the past.

**Comment by pstorch** [comment]
@Matze2 exactly. Do we have an issue for "can not clear coordinates in a waypoint"? I couldn't find it.

**Comment by Matze2** [comment]
I don't know. I added it to my waypoint editor issues list in #6823.
I can turn this list into an issue if you want.

**Comment by pstorch** [comment]
@Matze2 ok, I don't think we need another issue. Either we have to change it within this issue here or in #6823.

**Comment by S-Bartfast** [comment]
I think the 'invalid coordinates' can be dealt with here, but if you'd like your bearing/distance issue addressed it would be best to create a separate issue for that.

Note that I have now closed #6823 and created an new issue for combining the Input Coordinates Dialog and the Calculator into a single dialog (#6833).

**Comment by pstorch** [comment]
For me this issue is fixed. Invalid calculated coordinates are no longer visible in the waypoint list.
How to handle invalid values inside the calculator are handled in #6855 .
